### PR TITLE
handle CORE-3028

### PR DESCRIPTION
**Pull request by mmaxim**
@maxtaco @patrickxb 

Add parallelism to accessing the StorageMachine so we don't block the main loop every time we touch it. I'm going to get some tests going, but I wanted to get this in front you guys to make sure you generally agree with the direction I went in.

Thanks!

### CODE REVIEW

**Comment by mmaxim** [comment]
updated to remove `consumeCh` and `syncCh`

File: bin/vendor/github.com/keybase/gregor/rpc/server/main.go [hunk 1]
```diff
@@ -92,24 +92,31 @@ type Server struct {
...
-		broadcastTimeout: broadcastTimeout,
+		clock:             clockwork.NewRealClock(),
+		users:             make(map[string]*perUIDServer),
+		lastConns:         make(map[string]connectionID),
+		newConnectionCh:   make(chan *connection),
+		statsCh:           make(chan chan *Stats, 1),
+		broadcastCh:       make(chan messageArgs),
+		closeCh:           make(chan struct{}),
+		confirmCh:         make(chan confirmUIDShutdownArgs),
+		storageDispatchCh: make(chan storageReq, 10000),
```
Comment (maxtaco) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: reminder to make this tunable

File: rpc/server/main.go [hunk 2]
```diff
@@ -314,6 +320,71 @@ func (s *Server) Serve() error {
...
+	}
+	return nil
+}
+
+// storageDispatchHandler handles pulling requests off the storageDispatchCh queue
+// and routing them to the proper StorageMachine function. Many of these
+// run at once in different threads. This also makes it such that
+// any StorageMachine implementation must be thread safe.
+func (s *Server) storageDispatchHandler() {
+	for req := range s.storageDispatchCh {
```
Comment (maxtaco) [inline rpc/server/main.go]: is there also a need to listen on the end of the program too, and bail out then?  I've seen it in some of @patrickxb's similar handler loops.

Comment (mmaxim) [inline rpc/server/main.go]: Why?

Comment (mmaxim) [inline rpc/server/main.go]: I'll just add it.

File: rpc/server/main.go [hunk 3]
```diff
@@ -314,6 +320,71 @@ func (s *Server) Serve() error {
...
+
+type syncReq struct {
+	sm     gregor.StateMachine
+	log    rpc.LogOutput
+	arg    gregor1.SyncArg
+	respCh chan<- syncRet
+}
+
+// storageConsumeMessage schedules a Consume request on the dispatch handler
+func (s *Server) storageConsumeMessage(m gregor.Message, retCh chan<- error) {
```
Comment (maxtaco) [inline rpc/server/main.go]: make the `retCh` in this function and then just return `error` when done.

File: rpc/server/main.go [hunk 4]
```diff
@@ -314,6 +320,71 @@ func (s *Server) Serve() error {
...
+	err := s.storageDispatch(req)
+	if err != nil {
+		retCh <- err
+	}
+}
+
+// storageSync schedules a Sync request on the dispatch handler
+func (s *Server) storageSync(sm gregor.StateMachine, log rpc.LogOutput, arg gregor1.SyncArg,
+	retCh chan<- syncRet) {
+	req := syncReq{sm: sm, log: log, arg: arg, respCh: retCh}
```
Comment (maxtaco) [inline rpc/server/main.go]: same thing as above, don't pass in the channel if possible, just make it local to this function.

**Comment by maxtaco** [comment]
This looks good to me.  If you see a chance to simplify some of these methods, you should do it.   Also, please coordinate with @patrickxb about how to handle conflicts in the critiical serving path.

File: rpc/server/main.go (hunk 2 above)
Comment (mmaxim) [inline rpc/server/main.go]: I took @patrickxb's suggestion and just close the dispatch channel in `Shutdown`

**Comment by patrickxb** [comment]
LGTM

**Pull request merged**